folding-schemes/src/folding/nova/serialize.rs

```diff
@@ -0,0 +1,306 @@
+        self.i.serialize_with_mode(&mut writer, compress)?;
+        self.z_0.serialize_with_mode(&mut writer, compress)?;
+        self.z_i.serialize_with_mode(&mut writer, compress)?;
+        self.w_i.serialize_with_mode(&mut writer, compress)?;
+        self.u_i.serialize_with_mode(&mut writer, compress)?;
+        self.W_i.serialize_with_mode(&mut writer, compress)?;
+        self.U_i.serialize_with_mode(&mut writer, compress)?;
+        self.cf_W_i.serialize_with_mode(&mut writer, compress)?;
+        self.cf_U_i.serialize_with_mode(&mut writer, compress)?;
+        self.r1cs.serialize_with_mode(&mut writer, compress)?;
+        self.cf_r1cs.serialize_with_mode(&mut writer, compress)?;
+        // we serialize PoseidonConfig without implementing the `CanonicalDeserialize` trait
+        // this is because we can not implement an external trait on an external type.
+        self.poseidon_config
+            .full_rounds
+            .serialize_with_mode(&mut writer, compress)?;
+        self.poseidon_config
+            .partial_rounds
+            .serialize_with_mode(&mut writer, compress)?;
+        self.poseidon_config
+            .alpha
+            .serialize_with_mode(&mut writer, compress)?;
+        self.poseidon_config
+            .ark
+            .serialize_with_mode(&mut writer, compress)?;
+        self.poseidon_config
+            .mds
+            .serialize_with_mode(&mut writer, compress)?;
+        self.poseidon_config
+            .rate
+            .serialize_with_mode(&mut writer, compress)?;
+        self.poseidon_config
+            .capacity
+            .serialize_with_mode(&mut writer, compress)
+    }
+
+    fn serialized_size(&self, compress: ark_serialize::Compress) -> usize {
+        self.cs_params.serialized_size(compress)
+            + self.cf_cs_params.serialized_size(compress)
+            + self.i.serialized_size(compress)
+            + self.z_0.serialized_size(compress)
+            + self.z_i.serialized_size(compress)
+            + self.w_i.serialized_size(compress)
+            + self.u_i.serialized_size(compress)
+            + self.W_i.serialized_size(compress)
+            + self.U_i.serialized_size(compress)
+            + self.cf_W_i.serialized_size(compress)
+            + self.cf_U_i.serialized_size(compress)
+            + self.r1cs.serialized_size(compress)
+            + self.cf_r1cs.serialized_size(compress)
+            + self.poseidon_config.full_rounds.serialized_size(compress)
+            + self
+                .poseidon_config
+                .partial_rounds
+                .serialized_size(compress)
+            + self.poseidon_config.alpha.serialized_size(compress)
+            + self.poseidon_config.ark.serialized_size(compress)
+            + self.poseidon_config.mds.serialized_size(compress)
+            + self.poseidon_config.rate.serialized_size(compress)
+            + self.poseidon_config.capacity.serialized_size(compress)
```

Comment:
Sorry for the late comment :pray: , thought about it initially and discarded it, but now I'm giving it a second thought due the difference that it makes.

Do those params need to be serialized? The two R1CS, the Poseidon params, and the CommitmentScheme params.

They are quite big, and they can be regenerated without reading them from the serialized file. Also, they depend anyways on the definition of the configuration of the Nova struct (eg. even if importing the r1cs data, the Nova struct requires to have defined the FCircuit and internally will compute the r1cs for the AugmentedFCircuit).

What I'm thinking is that the
```
self.cs_params
self.cf_cs_params
self.r1cs
self.cf_r1cs
self.poseidon_config.full_rounds
self.poseidon_config.partial_rounds
self.poseidon_config.alpha
self.poseidon_config.ark
self.poseidon_config.mds
self.poseidon_config.rate
self.poseidon_config.capacity
```
parameters don't need to be serialized/deserialized, can be removed from the serialization, and can just be computed/taken as input in the deserialize method.


I've tested it locally, and the serialized data size went from `50 MB` to `6.4 MB` (`7.8x` reduction).

Reply:
I thought that it would be more user friendly this way. But indeed, this is almost a 10x reduction, which can be massive when using large circuits. I kept deriving the `Canonical(De)Serialize` traits on commitment schemes and `R1CS` since I think it might still be useful for users.

[d29bf0f](https://github.com/privacy-scaling-explorations/sonobe/pull/107/commits/d29bf0f493159bde40f6c09cbcd8b3435dc96c32)

Reply:
Nice, how would you see computing the r1cs data in the deserialize method too, avoiding the need for the user to obtain them outside?
The reasoning is that in the current test it's not a problem but because they already exist in the `nova` struct before serializing/deserializing, but in a real usage the dev deserializing back to Nova would not have at hand the r1cs data, and it's data that the lib can already obtain from the given FCircuit. 

It would be adding the lines from 300 to 315 from [nova/mod.rs#:300](https://github.com/privacy-scaling-explorations/sonobe/blob/5ea55cf54e1d1fc8ca523cab07f1c42fc6bade3f/folding-schemes/src/folding/nova/mod.rs#L300) into the `deserialize_nova` method and removing the inputs `r1cs` & `cf_r1cs`.

Reply:
Yup good idea, wil make things easier [73ce876](https://github.com/privacy-scaling-explorations/sonobe/pull/107/commits/73ce87634c17b608f5da099c776d3856de9cb45c)